rfcs/2020-07-28-3642-jmx_rfc.md

```diff
@@ -0,0 +1,239 @@
+# RFC 3642 - 2020-08-28 - Collecting metrics from the JVM via JMX
+
+This RFC is to introduce a new metrics source to consume metrics from Java Virtual Machines (JVM) using the JMX protocol. The high level plan is to implement one source that collects metrics from JVM servers.
+
+Background reading on JVM/JMX monitoring:
+
+- https://sysdig.com/blog/jmx-monitoring-custom-metrics/
+
+## Scope
+
+This RFC will cover:
+
+- A new source for JVM-based metrics using JMX.
+
+## Motivation
+
+Users want to collect, transform, and forward metrics to better observe how their JVM-based applications are performing.
+
+## Internal Proposal
+
+Build a single source called `jmx_metrics` (name to be confirmed) to collect JVM metrics.
+
+The recommended implementation is to use a/the Rust JMX client to connect to a JVM server by an address specified in configuration.
+
+- https://docs.rs/jmx/0.2.1/jmx/index.html
+- https://github.com/stefano-pogliani/jmx-rust
+
+This will require the end user to configure JMX on their JVM instances bound to an external port that can be queried by Vector.
+
+The Prometheus JMX exporter provides a Java-based agent claiming:
+
+```text
+This exporter is intended to be run as a Java Agent, exposing a HTTP server and serving metrics of the local JVM. It can be also run as an independent HTTP server and scrape remote JMX targets, but this has various disadvantages, such as being harder to configure and being unable to expose process metrics (e.g., memory and CPU usage). Running the exporter as a Java Agent is thus strongly encouraged.
+```
+
+We don't necessarily concur but it may require some testing as we proceed.
+
+The query will return these metrics by parsing the query results and converting them into metrics:
+
+- `jmx_up` -> Used as an uptime metric (0/1) ? - merits a broader discussion.
```

Comment:
:+1: I think these up/down metrics are useful. It lets you know if the process is even listening.